### ISSUE TITLE
NoViableAltException in a delete on a many-to-one id

### ISSUE THREAD

**Issue by mazharqayyum** [nhibernate/nhibernate-core#2688]
I am upgrading my application making use of Nhibernate 5.0.3 to Nhibernate 5.3.6. I had to update deprecated eager loads, use updated GetCurrentTransaction method and everything seems to work except one UNIT TEST. Here is the exception the code is throwing when it tries to run a HQL delete statement.

>  Antlr.Runtime.NoViableAltException : A recognition error occurred.
> 	at NHibernate.Hql.Ast.ANTLR.ErrorCounter.ThrowQueryException()
> 	at NHibernate.Hql.Ast.ANTLR.Exec.BasicExecutor..ctor(IStatement statement, IQueryable persister)
> 	at NHibernate.Hql.Ast.ANTLR.QueryTranslatorImpl.BuildAppropriateStatementExecutor(IStatement statement)
> 	at NHibernate.Hql.Ast.ANTLR.QueryTranslatorImpl.DoCompile(IDictionary`2 replacements, Boolean shallow, String collectionRole)
> 	at NHibernate.Hql.Ast.ANTLR.ASTQueryTranslatorFactory.CreateQueryTranslators(IQueryExpression queryExpression, IASTNode ast, String queryIdentifier, String collectionRole, Boolean shallow, IDictionary`2 filters, ISessionFactoryImplementor factory)
> 	at NHibernate.Hql.Ast.ANTLR.ASTQueryTranslatorFactory.CreateQueryTranslators(IQueryExpression queryExpression, String collectionRole, Boolean shallow, IDictionary`2 filters, ISessionFactoryImplementor factory)
> 	at NHibernate.Engine.Query.QueryPlanCache.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow, IDictionary`2 enabledFilters)
> 	at NHibernate.Impl.AbstractSessionImpl.GetHQLQueryPlan(IQueryExpression queryExpression, Boolean shallow)
> 	at NHibernate.Impl.AbstractSessionImpl.CreateQuery(String queryString)
> 	DomainModel\NHibernateHelper.cs(243,0): at CommerceBuilder.DomainModel.NHibernateHelper.CreateQuery(String query)
> 	Products\ProductVariantManager.cs(237,0): at CommerceBuilder.Products.ProductVariantManager.ResetVariantGrid(Int32 productId)
> 	Repositories\ProductOptionRepositoryTests.cs(58,0): at CommerceBuilder.Tests.Repositories.ProductOptionRepositoryTests.CanDelete()
> 	--NoViableAltException
> 	at NHibernate.Hql.Ast.ANTLR.SqlGenerator.whereClauseExpr()

The statement which is failing is this 
`("delete from BasketItem where Product.Id = " + productId).ExecuteUpdate()`

Product is mapped as many-to-on for BasketItem with not found mode set to ignore
`<many-to-one name="Product" column="ProductId" class="CommerceBuilder.Products.Product,CommerceBuilder" not-found="ignore" />`

Any idea what could be the problem?

**Comment by mazharqayyum** [comment]
Here is the debug log

> DEBUG 2021-02-25 14:27:28,456  6302ms QueryPlanCache         lambda_method      - unable to locate HQL query plan in cache; generating (delete from BasketItem where Product.Id = 1)
> DEBUG 2021-02-25 14:27:28,468  6314ms HqlParseEngine         lambda_method      - parse() - HQL: delete from BasketItem where Product.Id = 1
> DEBUG 2021-02-25 14:27:28,477  6323ms SessionImpl            lambda_method      - [session-id=ab216c20-18e6-4f0d-bb64-0f796bde6dad] executing real Dispose(False)
> DEBUG 2021-02-25 14:27:28,508  6354ms ErrorCounter           lambda_method      - throwQueryException() : no errors
> DEBUG 2021-02-25 14:27:28,530  6376ms HqlSqlWalker           lambda_method      - delete << begin [level=1, statement=delete]
> DEBUG 2021-02-25 14:27:28,537  6383ms FromElement            lambda_method      - FromClause{level=1} :  CommerceBuilder.Orders.BasketItem (no alias) -> basketitem0_
> INFO  2021-02-25 14:27:28,550  6396ms HqlSqlWalker           lambda_method      - attempting to resolve property [Product] as a non-qualified ref
> DEBUG 2021-02-25 14:27:28,551  6397ms FromReferenceNode      lambda_method      - Resolved :  {synthetic-alias} -> {synthetic-alias}
> INFO  2021-02-25 14:27:28,553  6399ms FromElement            lambda_method      - handling property dereference [CommerceBuilder.Orders.BasketItem () -> Product (Class)]
> DEBUG 2021-02-25 14:27:28,555  6401ms DotNode                lambda_method      - getDataType() : Product -> NHibernate.Type.ManyToOneType(CommerceBuilder.Products.Product)
> DEBUG 2021-02-25 14:27:28,557  6403ms DotNode                lambda_method      - dereferenceEntityJoin() : generating join for Product in CommerceBuilder.Orders.BasketItem {no alias} parent = [ ( . ( {non-qualified-property-ref} {synthetic-alias} Product ) Id ) ]
> INFO  2021-02-25 14:27:28,559  6405ms FromElementType        lambda_method      - Using non-qualified column reference [Product -> ([ProductId])]
> DEBUG 2021-02-25 14:27:28,560  6406ms FromElement            lambda_method      - FromClause{level=1} :  CommerceBuilder.Products.Product (no alias) -> product1_
> DEBUG 2021-02-25 14:27:28,561  6407ms FromClause             lambda_method      - addJoinByPathMap() : {synthetic-alias}.Product -> ac_Products product1_
> DEBUG 2021-02-25 14:27:28,561  6407ms FromElementFactory     lambda_method      - createEntityJoin() : Implied entity join
> DEBUG 2021-02-25 14:27:28,562  6408ms FromReferenceNode      lambda_method      - Resolved :  {synthetic-alias}.Product -> ProductId
> INFO  2021-02-25 14:27:28,562  6408ms FromElement            lambda_method      - handling property dereference [CommerceBuilder.Products.Product () -> Id (Class)]
> DEBUG 2021-02-25 14:27:28,562  6408ms DotNode                lambda_method      - getDataType() : Id -> NHibernate.Type.Int32Type
> INFO  2021-02-25 14:27:28,563  6409ms FromElementType        lambda_method      - Using non-qualified column reference [Id -> ([ProductId])]
> DEBUG 2021-02-25 14:27:28,563  6409ms FromReferenceNode      lambda_method      - Resolved :  {synthetic-alias}.Product.Id -> ProductId
> DEBUG 2021-02-25 14:27:28,567  6413ms HqlSqlWalker           lambda_method      - delete : finishing up [level=1, statement=delete]
> DEBUG 2021-02-25 14:27:28,573  6419ms JoinProcessor          lambda_method      - Using FROM fragment [ac_BasketItems]
> DEBUG 2021-02-25 14:27:28,575  6421ms JoinProcessor          lambda_method      - Using FROM fragment [ac_Products product1_]
> DEBUG 2021-02-25 14:27:28,575  6421ms SyntheticAndFactory    lambda_method      - Using unprocessed WHERE-fragment [ProductId=product1_.ProductId]
> DEBUG 2021-02-25 14:27:28,575  6421ms SyntheticAndFactory    lambda_method      - Using processed WHERE-fragment [ProductId=product1_.ProductId]
> DEBUG 2021-02-25 14:27:28,576  6422ms HqlSqlWalker           lambda_method      - delete >> end [level=1, statement=delete]
> DEBUG 2021-02-25 14:27:28,576  6422ms ErrorCounter           lambda_method      - throwQueryException() : no errors
> ERROR 2021-02-25 14:27:28,592  6438ms Parser                 lambda_method      - NoViableAltException(153@[])
> DEBUG 2021-02-25 14:27:28,593  6439ms ErrorCounter           lambda_method      - NoViableAltException(153@[])
> NoViableAltException(153@[])
> ERROR 2021-02-25 14:27:28,594  6440ms Parser                 lambda_method      - MismatchedTreeNodeException(39!=3)
> DEBUG 2021-02-25 14:27:28,594  6440ms ErrorCounter           lambda_method      - MismatchedTreeNodeException(39!=3)
> MismatchedTreeNodeException(39!=3)

**Comment by bahusoid** [comment]
Does it work with paramater?
```C#
("delete from BasketItem where Product.Id = :id" ).SetParamater("id", productId).ExecuteUpdate();
```

**Comment by fredericDelaporte** [comment]
This is a 5.3.x regression. I have added a test with 14d38d8, it passes on 5.2.x, but not on 5.3.x.
The test also passes if we remove the not-found setting.

**Comment by mazharqayyum** [comment]
Looks like we both were writing test cases at same time :) I sent a PR, let's see if it helps you in anyway.

**Comment by fredericDelaporte** [comment]
Fixed by #2691.
